pkg/vsphere/extraconfig/extraconfig_test.go

```diff
@@ -0,0 +1,584 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extraconfig
+
+import (
+	"net/url"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+// [BEGIN] SLIMMED DOWNED and MODIFIED VERSION of github.com/vmware/vic/metadata
+type Common struct {
+	ExecutionEnvironment string
+
+	ID string `vic:"0.1" scope:"hidden" key:"id"`
+
+	Name string `vic:"0.1" scope:"hidden" key:"name"`
+
+	Notes string `vic:"0.1" scope:"hidden" key:"notes"`
+}
+
+type ContainerVM struct {
+	Common `vic:"0.1" scope:"read-only" key:"common"`
+
+	Version string `vic:"0.1" scope:"hidden" key:"version"`
+
+	Aliases map[string]string
+
+	Interaction url.URL
+
+	AgentKey []byte
+}
+
+type ExecutorConfig struct {
+	Common `vic:"0.1" scope:"read-only" key:"common"`
+
+	Sessions map[string]SessionConfig `vic:"0.1" scope:"hidden" key:"sessions"`
+
+	Key []byte `json:"byte"`
+}
+
+type Cmd struct {
+	Path string `vic:"0.1" scope:"hidden" key:"path"`
+
+	Args []string `vic:"0.1" scope:"hidden" key:"args"`
+
+	Env []string `vic:"0.1" scope:"hidden" key:"env"`
+
+	Dir string `vic:"0.1" scope:"hidden" key:"dir"`
+
+	Cmd *exec.Cmd `vic:"0.1" scope:"hidden" key:"cmd"`
+}
+
+type SessionConfig struct {
+	Common `vic:"0.1" scope:"hidden" key:"common" json:"page"`
+
+	Cmd Cmd `vic:"0.1" scope:"hidden" key:"cmd"`
+
+	Tty bool `vic:"0.1" scope:"hidden" key:"tty"`
+}
+
+// [END] SLIMMED VERSION of github.com/vmware/vic/metadata
+
+func TestBasic(t *testing.T) {
+	type Type struct {
+		Int    int     `vic:"0.1" scope:"read-write" key:"int"`
+		Bool   bool    `vic:"0.1" scope:"read-write" key:"bool"`
+		Float  float64 `vic:"0.1" scope:"read-write" key:"float"`
+		String string  `vic:"0.1" scope:"read-write" key:"string"`
+	}
+
+	Struct := Type{
+		42,
+		true,
+		3.14,
+		"Grrr",
+	}
+
+	encoded := Encode(Struct)
+	expected := []types.BaseOptionValue{
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.int", Value: "42"},
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.bool", Value: "true"},
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.float", Value: "3.14E+00"},
+		&types.OptionValue{DynamicData: types.DynamicData{}, Key: "guestinfo.string", Value: "Grrr"},
```

Comment:
Yeah multiple guestinfo looks like a bug. I'm not sure why we want id as read-write in this example, Outside marks the whole struct read-only